dotcom-rendering/src/web/components/LiveEpic.importable.tsx

```diff
@@ -0,0 +1,288 @@
+
+	if (!consent) return;
+	log('dotcom', 'LiveEpic has consent state');
+	if (!countryCode) return;
+	log('dotcom', 'LiveEpic has countryCode');
+
+	return {
+		tracking: {
+			ophanPageId: window.guardian.config.ophan.pageViewId,
+			platformId: 'GUARDIAN_WEB',
+			clientName: 'dcr',
+			referrerUrl: window.location.origin + window.location.pathname,
+		},
+		targeting: {
+			contentType: 'LiveBlog',
+			sectionId: section,
+			shouldHideReaderRevenue: shouldHideReaderRevenue ?? false,
+			isMinuteArticle: true,
+			isPaidContent,
+			tags,
+			showSupportMessaging: !shouldHideSupportMessaging(isSignedIn),
+			isRecurringContributor: isRecurringContributor(isSignedIn),
+			lastOneOffContributionDate:
+				getLastOneOffContributionTimestamp() ?? undefined,
+			mvtId,
+			countryCode,
+			epicViewLog: storage.local.get('gu.contributions.views') || [],
+			weeklyArticleHistory:
+				storage.local.get('gu.history.weeklyArticleCount') || [],
+			hasOptedOutOfArticleCount: !consentGivenForArticleCounts(consent),
+			modulesVersion: 'v3',
+			url: window.location.origin + window.location.pathname,
+		},
+	};
+};
+
+/**
+ * Render
+ *
+ * Dynamically imports and renders a given module
+ *
+ * @param url string - The url of the compoenent
+ * @param name string - The name of the compoenent
+ * @param props object - The props of the compoenent
+ * @returns The resulting react component
+ */
+const Render = ({
+	url,
+	name,
+	props,
+}: {
+	url: string;
+	name: string;
+	props: { [key: string]: unknown };
+}) => {
+	// Using state here to store the Epic component that gets imported allows
+	// us to render it with React (instead of inserting it into the dom manually)
+	const [Epic, setEpic] = useState<React.FC<{ [key: string]: unknown }>>();
+
+	useEffect(() => {
+		setAutomat();
+		window
+			.guardianPolyfilledImport(url)
+			.then((epicModule) => {
+				// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+				setEpic(() => epicModule[name]); // useState requires functions to be wrapped
+			})
+			.catch((err) => {
+				const msg = `Error importing LiveBlog epic: ${err}`;
+				window.guardian.modules.sentry.reportError(
+					new Error(msg),
+					'liveblog-epic',
+				);
+			});
+	}, [url, name]);
+
+	if (!Epic) return null;
+	log('dotcom', 'LiveEpic has the Epic');
+
+	return (
+		<aside
+			css={css`
+				margin-bottom: ${space[3]}px;
+			`}
+		>
+			{/* eslint-disable-next-line react/jsx-props-no-spreading */}
+			<Epic {...props} />
+		</aside>
+	);
+};
+
+const Fetch = ({
+	payload,
+	contributionsServiceUrl,
+}: {
+	payload: unknown;
+	contributionsServiceUrl: string;
+}) => {
+	let url = joinUrl(contributionsServiceUrl, 'liveblog-epic');
+
+	// Check if we've been asked to force this epic
+	const params = new URLSearchParams(window.location.search);
+	if (params.get('force-liveblog-epic')) {
+		url += '?force=true';
+	}
+
+	// Send the payload to Contributions to request a module
+	const { data } = useApi<{
+		data: {
```

Comment:
It's a shame to lose the common [logic and types from SDC](https://github.com/guardian/support-dotcom-components/blob/main/packages/dotcom/src/requests.ts#L10) here.
I think we should have a common way to do these api requests for all RR channels (epic, liveblog epic, banner, header links), and DCR should not define the types.

Also, what are the benefits of `useApi` here? I'm not disputing its use, I just don't know enough about it.

Reply:
> It's a shame to lose the common [logic and types from SDC](https://github.com/guardian/support-dotcom-components/blob/main/packages/dotcom/src/requests.ts#L10) here. 

Agreed. I've made a change to import this type and use that instead

> I think we should have a common way to do these api requests for all RR channels (epic, liveblog epic, banner, header links)

Yeah, this makes a lot of sense. I really like how this PR is highlighting some interesting abstractions. `useEpic`, `usePayload`, etc. These could all - potentially - be hoisted up and used elsewhere. This fetch call could follow a similar route. I think it's good to be cautious though, the wrong abstraction done too early can be costly. But yeah, centralising this logic is a good shout.

> 
> Also, what are the benefits of `useApi` here? I'm not disputing its use, I just don't know enough about it.

We're in the process of moving all of our server state into [SWR](https://swr.vercel.app/). It's a fairly minimal library for storing state using React. Fetching data, storing it, passing it into components, deduplicating requests, caching, etc. is all tricky stuff to do and typically leads to complicated, brittle code. With SWR, you don't really need to think about it, you get automatic solutions for all sorts of common issues like these out the box.

It's obviously React focussed though and will not work at all outside of a React component.

`useApi` itself, is DCR's wrapper on top of SWR. It really just manages passing options and checking for errors.

In more general terms, we're keen to centralise data fetching on the platform. Not because we want to stop any other code making a fetch request but more because we want to make it easier to reason about where and when data is being brought into the stack. This is a medium term goal but one that we try to use to help made decisions. So here, where a fetch call is made which affects the main content on the page, ideally we would have more control over this at the platform level so that if we wanted to, say, introduce a loading state, then that's easier where it's the platform controlling the network call.

Reply:
Thanks, that all sounds great.

At some point it would be nice to have all of these SDC requests follow the same pattern. The others are:
- [banner](https://github.com/guardian/dotcom-rendering/blob/main/dotcom-rendering/src/web/components/StickyBottomBanner/ReaderRevenueBanner.tsx#L189)
- [epic](https://github.com/guardian/dotcom-rendering/blob/main/dotcom-rendering/src/web/components/SlotBodyEnd/ReaderRevenueEpic.tsx#L123)
- [header links](https://github.com/guardian/dotcom-rendering/blob/main/dotcom-rendering/src/web/components/ReaderRevenueLinks.tsx#L201)

Currently the `force-liveblog-epic` part comes for free with these existing calls

Reply:
100% agree. I also appreciate you don't want multiple ways to do the same thing to exist in parallel for a long time